File: src/main/java/ru/hh/typofixer/Morph.java
package ru.hh.typofixer;

import com.google.common.base.Function;
import com.google.common.collect.Iterables;

abstract public class Morph {
  public Token apply(Token token) {
    if (token instanceof Word) {
      return apply((Word) token);
    } else if (token instanceof Phrase) {
      return apply((Phrase) token);
    } else if (token instanceof Variant) {
      return apply((Variant) token);
    } else {
      throw new IllegalArgumentException(token.getClass().getCanonicalName());
    }
  }

  public Token apply(Word token) {
    return token;
  }

  public Variant apply(Variant token) {
    return new Variant(apply(token.getValues()));
  }

  public Phrase apply(Phrase token) {
    return new Phrase(apply(token.getValues()));
  }

  public Iterable<? extends Token> apply(Iterable<? extends  Token> tokens) {
    return Iterables.concat(
       Iterables.transform(tokens,
          new Function<Token, Token>() {
            @Override
            public Token apply(Token input) {
              return Morph.this.apply(input);
            }
          }));
  }
}

File: src/main/scala/ru/hh/typofixer/Morph.scala
package ru.hh.typofixer

import java.lang.IllegalArgumentException

abstract class Morph {
  def apply(token: Token): Token =
    token match {
      case token: Word => if (token.found) token else apply(token)
      case token: Phrase => apply(token)
      case token: Variant => apply(token)
      case _ => throw new IllegalArgumentException(token.getClass.getCanonicalName)
    }

  def apply(token: Word): Token = token

  def apply(token: Variant): Token = new Variant(token.values.map(apply))

  def apply(token: Phrase): Token = new Phrase(token.values.map(apply))

  def substrToWord(value: String, from: Int, to: Int): Word = {
    new Word(value.substring(from, to))
  }
}
